Fix #26387: Style dialog's left sidebar resizes unexpectedly when clicking certain pages

franciscolh04 opened the pull request:
Resolves: #26387

In the Style Dialog, clicking certain pages caused the left sidebar to resize unexpectedly.
This was due to a `connect` inside `EditStyle::adjustPagesStackSize()` that updated `setMinimumSize()` every time a new page was selected from the pageList, leading to unwanted horizontal shifts.
I removed this `connect`, which stopped the unexpected movement of the scroll bar.

However, I then noticed that some pages became unusable when the vertical scroll bar was dragged too far to the right because their buttons were positioned beyond the visible area of the dialog.
To fix this, I added QScrollAreas to those pages (*Glissandos & note-anchored lines*, *Fretboard diagrams*) to ensure all elements remained accessible.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually

franciscolh04 committed:
Fix #26387:left sidebar resizes unexpectedly when clicking certain pages

In the Style Dialog, clicking certain pages caused the left sidebar to
resize unexpectedly. This was due to a `connect` inside
`EditStyle::adjustPagesStackSize()` that updated `setMinimumSize()`
every time a new page was selected from the pageList, leading to
unwanted horizontal shifts. I removed this `connect`, which stopped
the unexpected movement of the scroll bar.
However, I then noticed that some pages became unusable when the
vertical scroll bar was dragged too far to the right because their
buttons were positioned beyond the visible area of the dialog.
To fix this, I added QScrollAreas to those pages
(*Glissandos & note-anchored lines*, *Fretboard diagrams*) to ensure
all elements remained accessible.

cbjeukendrup commented:
I see a number of changes that seem unnecessary, or at least their purpose isn’t completely clear. For example, some hard-coded minimum sizes, some changed size policies. Could you please double-check which ones are really necessary?
Tip: use GitHub’s “hide whitespace changes” feature to make the diff view more understandable.

franciscolh04 commented:
@cbjeukendrup, thanks for the feedback!
I used Qt Creator to make some adjustments, so that might have introduced a few unintended changes.
Thanks for the tip about hiding whitespace changes — that definitely helps!
I'll review the changes and get back to you shortly.

franciscolh04 committed:
Fix: Revert unnecessary size policy and geometry property changes in Edit Style dialog

Restored original hardcoded geometry values (98x28) and a size policy value that didn't affect the layout.
Related to #26387.

franciscolh04 commented:
Hi!
I just noticed that one of the tests failed — I’m guessing I need to rebase onto the latest `master`, right?

franciscolh04 commented:
The rebase solved the problem of the tests, thanks!

franciscolh04 commented:
Hey @cbjeukendrup ! Is there any chance I get some more feedback on the PR? If there is anything wrong let me know and I'll get back to it!

cbjeukendrup commented on src/notation/view/widgets/editstyle.cpp:
```diff
@@ -1467,14 +1467,6 @@ void EditStyle::adjustPagesStackSize(int currentPageIndex)
     QSize preferredSize = pageStack->widget(currentPageIndex)->sizeHint();
     pageStack->setMinimumSize(preferredSize);
-
-    connect(pageStack, &QStackedWidget::currentChanged, this, [this](int currentIndex) {
```
I'm wondering whether this removal is still necessary after adding the QScrollArea (because the scroll area ensures that `currentPage->sizeHint()` won't be so big that the sidebar will resize)

franciscolh04 replied:
Hey @cbjeukendrup! Thanks again for the feedback.
I double-checked on my setup, and removing the connect still seems necessary to properly fix the left sidebar’s resizing bug — the issue happened even on pages that already had a `QScrollArea`. The extra `QScrollAreas` were just to keep some specific pages fully usable after the fix.

cbjeukendrup commented on src/notation/view/widgets/editstyle.ui:
```diff
@@ -10164,31 +10164,74 @@
+           <layout class="QVBoxLayout" name="verticalLayout_59">
+            <item>
+             <widget class="QGroupBox" name="groupBox_glissando">
+              <property name="sizePolicy">
+               <sizepolicy hsizetype="Fixed" vsizetype="Preferred">
+                <horstretch>0</horstretch>
+                <verstretch>0</verstretch>
+               </sizepolicy>
+              </property>
+              <property name="minimumSize">
+               <size>
+                <width>627</width>
+                <height>0</height>
+               </size>
+              </property>
```
Why is this necessary?

franciscolh04 replied:
These sizePolicy and minimumSize values are necessary to ensure the layout behaves correctly. After testing several approaches, I found that these specific settings were the minimum required to make the design function as expected. They are directly related to the QScrollAreas I added, which are essential for ensuring the horizontal scroll bar on each page works properly and the content is displayed correctly. Without these settings, the content would overflow, or the horizontal scroll bar wouldn't appear when needed.

I understand that it's generally best to avoid hardcoded values, but after trying different solutions, I couldn't find another way to make the layout work as intended. 

If you'd prefer me to explore an alternative solution, please let me know! 🙂

cbjeukendrup commented on src/notation/view/widgets/editstyle.ui:
```diff
@@ -10164,31 +10164,74 @@
+            </item>
+            <item>
+             <widget class="QGroupBox" name="groupBox_noteline">
+              <property name="sizePolicy">
+               <sizepolicy hsizetype="Fixed" vsizetype="Preferred">
+                <horstretch>0</horstretch>
+                <verstretch>0</verstretch>
+               </sizepolicy>
+              </property>
+              <property name="minimumSize">
+               <size>
+                <width>627</width>
+                <height>0</height>
+               </size>
+              </property>
```
Same here

cbjeukendrup commented on src/notation/view/widgets/editstyle.ui:
```diff
@@ -14159,15 +14202,21 @@ first note of the system</string>
              <widget class="QWidget" name="fretboardsWidget" native="true">
               <property name="sizePolicy">
-               <sizepolicy hsizetype="Expanding" vsizetype="Expanding">
+               <sizepolicy hsizetype="Fixed" vsizetype="Expanding">
                 <horstretch>0</horstretch>
                 <verstretch>0</verstretch>
                </sizepolicy>
               </property>
+              <property name="minimumSize">
+               <size>
+                <width>470</width>
+                <height>0</height>
+               </size>
+              </property>
```
Why are these necessary?

cbjeukendrup commented:
I have to investigate the issue more myself, to check that the solution from this PR is the best one. But until the summer holidays I'm very busy still, so I don't have time to do it very soon. However, I'm not forgetting about it!

mathesoncalum commented:
Apologies for the radio silence @franciscolh04 - we appreciate your efforts on this one.

We're actually planning an overhaul to Style in version 5.0 (a major redesign and migrating to QML instead of Widgets) so it's unlikely we'll need these changes. I'm going to close this for now, also archiving for future reference in case there's a change of plans. Cheers!
